### PR TITLE
Force unicode before interaction

### PR DESCRIPTION
**Pull request by joguSD**
Prompt toolkit requires that the strings given to it are unicode. Data given to interactions can come from arbitrary sources (service calls) so there must be some mechanism in place ensuring that all strings given to prompt toolkit are of `six.text_type`. This function will recursively search lists and dicts for strings and encode any non-unicode strings as unicode. This is not ideal but is necessary to move forward.

@JordonPhillips @kyleknap

### CODE REVIEW

File: awsshell/utils.py [hunk 1]
```diff
@@ -118,3 +119,16 @@ def file_contents(self, filename, binary=False):
 
     def file_exists(self, filename):
         return filename in self._file_mapping
+
+
+def force_unicode(obj, encoding='utf8'):
```
Comment (kyleknap) [inline awsshell/utils.py]: It probably would be nice if you could hook into the botocore response parsers to do this but this looks like the next best thing as there is not really a way to do this.

File: awsshell/utils.py [hunk 2]
```diff
@@ -118,3 +119,16 @@ def file_contents(self, filename, binary=False):
...
+    """Recursively search lists and dicts for strings to encode as unicode."""
+    if isinstance(obj, dict):
+        for key in obj:
+            obj[key] = force_unicode(obj[key])
+    elif isinstance(obj, list):
+        obj = [force_unicode(val) for val in obj]
+    elif isinstance(obj, six.string_types):
```
Comment (kyleknap) [inline awsshell/utils.py]: How is binary handled by the shell? I am not too familiar with how the shell handles binary in terms of: Is binary base64 encoded in responses? If it is not base64 encoded, how do you handle the Python 2 case where the data is binary (i.e. it is considered a ``str``, allowing it to pass the ``six.sting_types`` and ``not six.text_type`` case) but cannot be utf-8 encoded as it is literally binary with no encoding?

File: tests/unit/test_utils.py [hunk 3]
```diff
@@ -93,3 +96,13 @@ def test_can_open_in_read(self):
             f.seek(0)
             assert f.read() == "foobar"
         self.assertFalse(os.path.isfile(filename))
+
+
+strings = ['', u'', 'some text', u'some text', 'yu\u3086', u'yu\u3086']
+
+
+@pytest.mark.parametrize('text', strings)
```
Comment (kyleknap) [inline tests/unit/test_utils.py]: It would probably be good to test more than just strings. Like dictionaries and lists as well to cover all of your bases.

**Comment by kyleknap** [comment]
Looks good. Just had comments about adding more test cases and a smaller possible edge case.

File: awsshell/utils.py (hunk 2 above)
Comment (joguSD) [inline awsshell/utils.py]: It would fail the encoding, causing an exception. I'll wrap that part in a try catch so that worst case is strings that are not utf8 compatible will not be converted.

File: awsshell/utils.py [hunk 4]
```diff
@@ -123,12 +125,19 @@ def file_exists(self, filename):
 
 def force_unicode(obj, encoding='utf8'):
     """Recursively search lists and dicts for strings to encode as unicode."""
+    def attempt_decode(string):
```
Comment (kyleknap) [inline awsshell/utils.py]: It is a little strange to define the function each time ``force_unicode`` is called. Typically I would extract this out to the module level or not even create a function in the first place since it is only used once.

File: tests/unit/test_utils.py [hunk 5]
```diff
@@ -106,3 +106,18 @@ def test_force_unicode(text):
...
+        'a': ['string'],
+        'b': {'str': 'yu\u3086'},
+        'c': '\x81',
+    }
+
+    clean_obj = force_unicode(obj)
+    assert isinstance(clean_obj['a'][0], six.text_type)
+    assert isinstance(clean_obj['b']['str'], six.text_type)
+    # c should fail decoding and remain unchanged
+    assert clean_obj['c'] is obj['c']
```
Comment (kyleknap) [inline tests/unit/test_utils.py]: I would make the unicode decoding issue a separate test as it does not really have anything to do with recursing through a list or dictionary.

File: awsshell/utils.py (hunk 4 above)
Comment (joguSD) [inline awsshell/utils.py]: Good catch, I think generally this isn't a super big deal but since this is a recursive function the redefinition could be pretty expensive.

**Comment by kyleknap** [comment]
Looks good. :ship:
